### PR TITLE
cleanup: fix typings for resolver.py

### PR DESCRIPTION
**Pull request by shubhbapna**
part of #226 

fixes:
```
src/fromager/resolver.py:79: error: Need type annotation for "tags" (hint: "tags: set[<type>] = ...")  [var-annotated]
src/fromager/resolver.py:82: error: Incompatible types in assignment (expression has type "frozenset[Tag]", variable has type "set[Any]")  [assignment]
src/fromager/resolver.py:150: error: Incompatible return value type (got "tuple[str, ...]", expected "tuple[str]")  [return-value]
src/fromager/resolver.py:150: error: Argument 1 to "sorted" has incompatible type "Iterable[str] | None"; expected "Iterable[str]"  [arg-type]
src/fromager/resolver.py:155: error: Function is missing a type annotation  [no-untyped-def]
src/fromager/resolver.py:155: error: Signature of "get_preference" incompatible with supertype "AbstractProvider"  [override]
src/fromager/resolver.py:155: note:      Superclass:
src/fromager/resolver.py:155: note:          def get_preference(self, identifier: str, resolutions: Mapping[str, Candidate], candidates: Mapping[str, Iterator[Candidate]], information: Mapping[str, Iterator[RequirementInformation[Requirement, Candidate]]], backtrack_causes: Sequence[RequirementInformation[Requirement, Candidate]]) -> Preference
src/fromager/resolver.py:155: note:      Subclass:
src/fromager/resolver.py:155: note:          def get_preference(self, identifier: Any, resolutions: Any, candidates: Any, information: Any, **kwds: Any) -> Any
src/fromager/resolver.py:166: error: Missing type parameters for generic type "list"  [type-arg]
src/fromager/resolver.py:173: error: Argument 2 of "find_matches" is incompatible with supertype "AbstractProvider"; supertype defines the argument type as "Mapping[str, Iterator[Requirement]]"  [override]
src/fromager/resolver.py:173: note: This violates the Liskov substitution principle
src/fromager/resolver.py:173: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
src/fromager/resolver.py:174: error: Argument 3 of "find_matches" is incompatible with supertype "AbstractProvider"; supertype defines the argument type as "Mapping[str, Iterator[Candidate]]"  [override]
src/fromager/resolver.py:174: note: This violates the Liskov substitution principle
src/fromager/resolver.py:174: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
src/fromager/resolver.py:196: error: Argument 2 of "find_matches" is incompatible with supertype "AbstractProvider"; supertype defines the argument type as "Mapping[str, Iterator[Requirement]]"  [override]
src/fromager/resolver.py:196: note: This violates the Liskov substitution principle
src/fromager/resolver.py:196: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
src/fromager/resolver.py:197: error: Argument 3 of "find_matches" is incompatible with supertype "AbstractProvider"; supertype defines the argument type as "Mapping[str, Iterator[Candidate]]"  [override]
src/fromager/resolver.py:197: note: This violates the Liskov substitution principle
src/fromager/resolver.py:197: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
src/fromager/resolver.py:199: error: Incompatible types in assignment (expression has type "list[Requirement]", variable has type "dict[str, Iterable[Requirement]]")  [assignment]
src/fromager/resolver.py:207: error: Argument 2 to "get_project_from_pypi" has incompatible type "set[Never]"; expected "tuple[str]"  [arg-type]
src/fromager/resolver.py:217: error: "str" has no attribute "specifier"  [attr-defined]
src/fromager/resolver.py:246: error: Argument 1 to "sorted" has incompatible type "list[Candidate]"; expected "Iterable[Version]"  [arg-type]
src/fromager/resolver.py:262: error: Argument 2 of "find_matches" is incompatible with supertype "AbstractProvider"; supertype defines the argument type as "Mapping[str, Iterator[Requirement]]"  [override]
src/fromager/resolver.py:262: note: This violates the Liskov substitution principle
src/fromager/resolver.py:262: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
src/fromager/resolver.py:263: error: Argument 3 of "find_matches" is incompatible with supertype "AbstractProvider"; supertype defines the argument type as "Mapping[str, Iterator[Candidate]]"  [override]
src/fromager/resolver.py:263: note: This violates the Liskov substitution principle
src/fromager/resolver.py:263: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
src/fromager/resolver.py:265: error: Incompatible types in assignment (expression has type "list[Requirement]", variable has type "dict[str, Iterable[Requirement]]")  [assignment]
src/fromager/resolver.py:291: error: "str" has no attribute "specifier"  [attr-defined]
src/fromager/resolver.py:305: error: Argument "url" to "Candidate" has incompatible type "str | Version"; expected "str"  [arg-type]
src/fromager/resolver.py:306: error: Argument 1 to "sorted" has incompatible type "list[Candidate]"; expected "Iterable[Version]"  [arg-type]
```

### CODE REVIEW

File: src/fromager/resolver.py [hunk 1]
```diff
@@ -36,7 +37,7 @@
 
 def get_project_from_pypi(
     project: str,
-    extras: tuple[str],
+    extras: tuple[str] | set[str],
```
Comment (tiran) [inline src/fromager/resolver.py:40]: `tuple[str]` is a tuple with one item. Candidate expects an `Iterable[str]`.

```suggestion
    extras: typing.Iterable[str],
```

File: src/fromager/resolver.py [hunk 2]
```diff
@@ -145,14 +146,27 @@ def identify(self, requirement_or_candidate: Requirement | Candidate) -> str:
     def get_extras_for(
         self,
         requirement_or_candidate: Requirement | Candidate,
-    ) -> tuple[str]:
+    ) -> tuple[str, ...]:
         # Extras is a set, which is not hashable
-        return tuple(sorted(requirement_or_candidate.extras))
+        if requirement_or_candidate.extras:
+            return tuple(sorted(list(requirement_or_candidate.extras)))
```
Comment (tiran) [inline src/fromager/resolver.py:152]: `sorted()` works with any iterable

```suggestion
            return tuple(sorted(requirement_or_candidate.extras))
```

File: src/fromager/resolver.py [hunk 3]
```diff
@@ -145,14 +146,27 @@ def identify(self, requirement_or_candidate: Requirement | Candidate) -> str:
...
 
     def get_base_requirement(self, candidate: Candidate) -> Requirement:
         return Requirement(f"{candidate.name}=={candidate.version}")
 
-    def get_preference(self, identifier, resolutions, candidates, information, **kwds):
+    def get_preference(
+        self,
+        identifier: str,
+        resolutions: typing.Mapping[str, Candidate],
+        candidates: typing.Mapping[str, typing.Iterator[Candidate]],
```
Comment (tiran) [inline src/fromager/resolver.py:162]: Why iterator? Could we use the CandidateMap type alias here?
```suggestion
        candidates: CandidateMap,
```

File: src/fromager/resolver.py (hunk 2 above)
Comment (shubhbapna) [reply src/fromager/resolver.py:152]: the `Requirement` type defines `extras` as a set which needs to be converted to a list

File: src/fromager/resolver.py (hunk 3 above)
Comment (shubhbapna) [reply src/fromager/resolver.py]: I had just copied these types from resolvelib

File: pyproject.toml [hunk 4]
```diff
@@ -105,11 +105,9 @@ mypy_path = ["src"]
 # TODO: remove excludes and silent follow
 follow_imports = "silent"
 exclude = [
-  "^src/fromager/resolver\\.py$",
   "^src/fromager/sources\\.py$",
   "^src/fromager/sdist\\.py$",
   "^src/fromager/commands/build\\.py$",
-  "^src/fromager/settings\\.py$",
```
Comment (dhellmann) [inline pyproject.toml:112]: mypy is failing for settings.py.

Comment (dhellmann) [inline pyproject.toml:112]: mypy is failing for settings.py.

Comment (shubhbapna) [reply pyproject.toml]: fixed!

**Review by dhellmann — APPROVED** [review]

**Pull request merged**